src/frontend/src/hooks/useExcelImport.ts

```diff
@@ -0,0 +1,119 @@
+import { useState, useCallback, useMemo, useEffect } from 'react';
+import { useAppContext } from '../appContext';
+import ExcelParseReponse, { ExcelHeader } from '../models/ExcelParseResponse';
+
+export type AutoGenerateColumn<T> = {
+    title: keyof T;
+    format: (columns: Column<T>[]) => {};
+};
+
+export type Column<T> = {
+    title: keyof T;
+    variations?: string[];
+    format?: (item: string) => {};
+};
+
+export type ExcelImportSettings<T> = {
+    columns: Column<T>[];
+    autoGenerateColumns?: AutoGenerateColumn<T>[];
+};
+
+type ColumnIndex<T> = {
+    [key in keyof T]: number;
+};
+
+const useExcelImport = <T>(excelImportSettings: ExcelImportSettings<T>) => {
+    const { apiClient } = useAppContext();
+
+    const [selectedFile, setSelectedFile] = useState<File | null>(null);
+    const [isProccessingFile, setIsProccessingFile] = useState<boolean>(false);
+    const [processedFile, setProcessedFile] = useState<T[] | null>(null);
+    const [processingError, setProcessingError] = useState<boolean>(false);
+    const { columns, autoGenerateColumns } = excelImportSettings;
+
+    useEffect(() => {
+        if (selectedFile === null) return;
+
+        processExcelFileAsync(selectedFile);
+    }, [selectedFile]);
+
+    const processExcelFileAsync = async (file: File) => {
+        setIsProccessingFile(true);
+        setProcessingError(false);
+
+        try {
+            const excelReponse = await apiClient.parseExcelFileAsync(file);
+
+            if (excelReponse) {
+                const formattedResponse = formatExcelReponse(excelReponse);
+                setProcessedFile(formattedResponse);
+            } else {
+                setProcessedFile(null);
+            }
+        } catch (e) {
+            setProcessingError(true);
+        } finally {
+            setIsProccessingFile(false);
+        }
+    };
+
+    const mapHeaderIndexesToColumns = (headers: ExcelHeader[]): ColumnIndex<T> => {
+        let columnIndexes = {};
+
+        for (const column of columns) {
+            const header = headers.find((h) =>
+                column.variations
+                    ? column.variations.includes(h.title.toLocaleLowerCase())
+                    : h.title.toLocaleLowerCase() === column.title.toString().toLocaleLowerCase()
+            );
+
+            columnIndexes = {
+                ...columnIndexes,
+                [column.title]: header?.colIndex || header?.colIndex === 0 ? header.colIndex : -1,
+            };
+        }
+
```

Comment:
Consider using Prototype.reduce() for these kind of operations to avoid variables, would something like this work?
```suggestion
        const columnIndexes = columns.reduce((indexes, column) => {
           const header = headers.find((h) =>
                column.variations
                    ? column.variations.includes(h.title.toLocaleLowerCase())
                    : h.title.toLocaleLowerCase() === column.title.toString().toLocaleLowerCase()
               );
               return {
                ...indexes,
                [column.title]: header?.colIndex || header?.colIndex === 0 ? header.colIndex : -1,
               },{})
        }
```